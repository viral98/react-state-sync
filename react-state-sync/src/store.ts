import createStore from './createStore'

const store = createStore({
  value1: 0,
  value2: 0
})

export type ValuesStore = ReturnType<typeof store.getState>

export default store

const useStore = (selector: (state: ValuesStore) => number) => {}

// eslint-disable-next-line react-hooks/rules-of-hooks
useStore(() => 2)
