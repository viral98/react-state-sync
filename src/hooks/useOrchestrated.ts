import createStore, { DefaultObject } from '../createStore'
import { BaseCacheResource } from '../resources/BaseCacheResource'
import { BaseResource } from '../resources/BaseResource'
import { BaseCacheService } from '../services/BaseCacheService'
import api from '../utils/api'

interface UseOrchestratedProps<T> {
  pathName: string
}
export function useOrchestrated<T extends DefaultObject>({ pathName }: UseOrchestratedProps<T>) {
  const concreteApi = api({})
  const concreteCacheService = new BaseCacheService<T>(pathName, concreteApi)
  const concreteBaseResource = new BaseResource<T>(concreteCacheService, pathName)

  return concreteBaseResource
}
