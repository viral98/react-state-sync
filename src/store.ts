import createStore from './createStore'
import { Book } from './types/books'

const store = createStore<Book>([])

export type ValuesStore = ReturnType<typeof store.getState>

export default store
