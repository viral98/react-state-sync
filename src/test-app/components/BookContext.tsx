import React from 'react'

const BookContext = React.createContext({
  Title: ''
})

export default BookContext
