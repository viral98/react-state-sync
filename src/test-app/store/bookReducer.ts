import { ActionTypes } from '../../actions/BaseActions'
import { DefaultObject, StoreState } from '../../createStore'
import { Action } from '../../types/action'
import { Book } from '../../types/books'

interface UpdateAction<Book> extends Action<ActionTypes.UPDATE> {
  payload: StoreState<Book>
}
interface DeleteAction<Book> extends Action<ActionTypes.DELETE> {
  payload: StoreState<Book>
}
interface GetAllAction<Book> extends Action<ActionTypes.GET_ALL> {
  payload: StoreState<Book[]>
}
interface PostAction<Book> extends Action<ActionTypes.POST> {
  payload: StoreState<Book>
}

interface ActionReturnType {
  type: string
  payload: StoreState<Book>
}

export type BaseActions<Book> =
  | UpdateAction<Book>
  | DeleteAction<Book>
  | GetAllAction<Book>
  | PostAction<Book>

export function UpdateValueInStore<Book extends DefaultObject>({
  payload
}: {
  payload: StoreState<Book>
}): ActionReturnType {
  return {
    type: ActionTypes.UPDATE,
    payload
  }
}

export function DeleteValueFromStore<Book extends DefaultObject>({
  payload
}: {
  payload: StoreState<Book>
}): ActionReturnType {
  return {
    type: ActionTypes.DELETE,
    payload
  }
}

export function PutAllValuesInStore<Book extends DefaultObject>({
  payload
}: {
  payload: StoreState<Book>
}): ActionReturnType {
  return {
    type: ActionTypes.GET_ALL,
    payload
  }
}

export function AddANewValueInStore<Book extends DefaultObject>({
  payload
}: {
  payload: StoreState<Book>
}): ActionReturnType {
  return {
    type: ActionTypes.POST,
    payload
  }
}

export function bookReducer<Book extends DefaultObject>(
  state: StoreState<Book>[],
  action: BaseActions<Book>
): StoreState<Book>[] {
  switch (action.type) {
    case ActionTypes.GET_ALL:
      return state

    case ActionTypes.UPDATE:
      return state.map((stateItem: StoreState<Book>) => {
        if (stateItem._id === action.payload._id) {
          return action.payload
        } else {
          return stateItem
        }
      })

    case ActionTypes.POST: {
      return [...state, action.payload]
    }
    case ActionTypes.DELETE:
      return state.filter((obj) => obj._id !== action.payload._id)

    default:
      console.error('Action must be a predefined actionType.')
  }
  return state
}
