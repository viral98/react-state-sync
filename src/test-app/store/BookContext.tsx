import React from 'react';
import { dummyBookResponse } from '../../hooks/testUtil';

export const BookContext = React.createContext(dummyBookResponse);
