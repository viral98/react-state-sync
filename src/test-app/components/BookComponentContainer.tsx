import React, { useContext, useEffect, useRef, useState } from 'react';
import { Book } from '../../types/books';
import { BookContext } from '../store/BookContext';
import BookComponentContext from './BookComponentContext';

function BookComponentContainer(): JSX.Element {
  const count = useRef(1)

  const [title, setTitle] = useState<string[]>([]);
  const books = useContext(BookContext);

  useEffect(() => {
    count.current += 1
  }, [])

  useEffect(() => {
    if(books && Array.isArray(books)){
      setTitle(books.map((book) => book.title as string));
    }
  }, []);

  return (
    <BookComponentContext selectedValues={title[0]}/>
  )
}

export default BookComponentContainer;
