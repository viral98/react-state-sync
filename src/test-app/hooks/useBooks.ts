import { Dispatch, useEffect, useReducer } from 'react'
import { act } from 'react-dom/test-utils'
import { ActionTypes } from '../../actions/BaseActions'
import { DefaultObject } from '../../createStore'
import { Book } from '../../types/books'
import { BaseActions, bookReducer } from '../store/bookReducer'

export const useBooks = (): {
  data: Book[] | null
  bookDispatch: Dispatch<BaseActions<DefaultObject>>
} => {
  const [bookState, bookDispatch] = useReducer(bookReducer, [])

  useEffect(() => {
    async function fetchBooks() {
      const response = await fetch('https://react-state-sync-serverless.vercel.app/api/books')
      const data = (await response.json()).data

      act(() => {
        bookDispatch({
          type: ActionTypes.GET_ALL,
          payload: data
        })
      })
    }

    fetchBooks()
  }, [])

  return {
    data: bookState,
    bookDispatch
  }
}
