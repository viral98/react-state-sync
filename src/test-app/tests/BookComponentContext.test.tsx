
import { render } from '@testing-library/react';
import React from 'react';
import { BookContext } from '../store/BookContext';
import BookComponentContainer from '../components/BookComponentContainer';
import { dummyBookResponse } from '../../hooks/testUtil';
import renderer from 'react-test-renderer';


function TestComponent() {
  return (
    <>
        <BookContext.Provider value={dummyBookResponse}>
            <BookComponentContainer />
        </BookContext.Provider>
    </>
  )
}

describe('Test', () => {

  it('renders the component correctly.', () => {
    const { getByTestId } = render(<TestComponent />);

    expect(getByTestId('count').textContent).toBe('2');

    const tree = renderer
      .create(<TestComponent />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders the component correctly.', () => {
    const tree = renderer
      .create(<TestComponent />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

});
