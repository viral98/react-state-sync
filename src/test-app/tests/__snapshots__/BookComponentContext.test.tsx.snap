// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test renders the component correctly. 1`] = `
[
  <div
    data-testid="count"
  >
    1
  </div>,
  <div
    data-testid="title"
  >
     
     
  </div>,
]
`;

exports[`Test renders the component correctly. 2`] = `
[
  <div
    data-testid="count"
  >
    2
  </div>,
  <div
    data-testid="title"
  >
     
    Text 2
     
  </div>,
]
`;
